File: sections/02_simulation_bruteforce/problems/c_best_subset/solution.py
"""Reference solution for C. Best Subset."""

import sys


def best_subset(values: list[int], limit: int) -> int:
    n = len(values)
    best = 0
    for mask in range(1 << n):
        total = 0
        for i, value in enumerate(values):
            if mask & (1 << i):
                total += value
        if total <= limit and total > best:
            best = total
    return best


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    t = data[0]
    idx = 1
    out: list[str] = []
    for _ in range(t):
        n = data[idx]
        limit = data[idx + 1]
        idx += 2
        values = data[idx:idx + n]
        idx += n
        out.append(str(best_subset(values, limit)))
    sys.stdout.write("\n".join(out))


if __name__ == "__main__":
    main()

File: sections/02_simulation_bruteforce/problems/c_best_subset/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int T;
    cin >> T;
    while (T--) {
        int n;
        long long limit;
        cin >> n >> limit;
        vector<long long> values(n);
        for (long long &value : values) {
            cin >> value;
        }

        long long best = 0;
        for (int mask = 0; mask < (1 << n); ++mask) {
            long long total = 0;
            for (int i = 0; i < n; ++i) {
                if (mask & (1 << i)) {
                    total += values[i];
                }
            }
            if (total <= limit) {
                best = max(best, total);
            }
        }
        cout << best << '\n';
    }

    return 0;
}

